File: Day-1/url-shortener-sample/java/URLShortener.java
import java.util.HashMap;
import java.util.Map;

public class URLShortener {
    private static final String ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

    private final String baseUrl;
    private final Map<String, String> codeToUrl;
    private final Map<String, String> urlToCode;
    private long counter;

    public URLShortener(String baseUrl) {
        this.baseUrl = baseUrl;
        this.codeToUrl = new HashMap<>();
        this.urlToCode = new HashMap<>();
        this.counter = 1L;
    }

    public synchronized String shortenUrl(String longUrl) {
        if (longUrl == null || !(longUrl.startsWith("http://") || longUrl.startsWith("https://"))) {
            throw new IllegalArgumentException("URL must start with http:// or https://");
        }

        if (urlToCode.containsKey(longUrl)) {
            return baseUrl + urlToCode.get(longUrl);
        }

        String code = encodeBase62(counter++);
        codeToUrl.put(code, longUrl);
        urlToCode.put(longUrl, code);
        return baseUrl + code;
    }

    public synchronized String resolve(String shortUrlOrCode) {
        String code = shortUrlOrCode;
        int slash = shortUrlOrCode.lastIndexOf('/');
        if (slash >= 0 && slash < shortUrlOrCode.length() - 1) {
            code = shortUrlOrCode.substring(slash + 1);
        }
        return codeToUrl.get(code);
    }

    private String encodeBase62(long n) {
        if (n == 0) {
            return String.valueOf(ALPHABET.charAt(0));
        }

        StringBuilder sb = new StringBuilder();
        int base = ALPHABET.length();
        long value = n;

        while (value > 0) {
            int rem = (int) (value % base);
            sb.append(ALPHABET.charAt(rem));
            value /= base;
        }

        return sb.reverse().toString();
    }
}

File: Day-1/url-shortener-sample/python/url_shortener.py
from __future__ import annotations

from dataclasses import dataclass, field
from threading import Lock


@dataclass
class URLShortener:
    base_url: str = "http://short.ly/"
    _code_to_url: dict[str, str] = field(default_factory=dict, init=False)
    _url_to_code: dict[str, str] = field(default_factory=dict, init=False)
    _counter: int = field(default=1, init=False)
    _lock: Lock = field(default_factory=Lock, init=False)

    _alphabet = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"

    def shorten_url(self, long_url: str) -> str:
        if not long_url or not long_url.startswith(("http://", "https://")):
            raise ValueError("URL must start with http:// or https://")

        with self._lock:
            existing_code = self._url_to_code.get(long_url)
            if existing_code:
                return self.base_url + existing_code

            code = self._encode_base62(self._counter)
            self._counter += 1

            self._code_to_url[code] = long_url
            self._url_to_code[long_url] = code
            return self.base_url + code

    def resolve(self, short_url_or_code: str) -> str | None:
        code = short_url_or_code.rsplit("/", maxsplit=1)[-1]
        return self._code_to_url.get(code)

    def _encode_base62(self, n: int) -> str:
        if n == 0:
            return self._alphabet[0]

        chars = []
        base = len(self._alphabet)
        while n > 0:
            n, rem = divmod(n, base)
            chars.append(self._alphabet[rem])
        chars.reverse()
        return "".join(chars)
